src/Microsoft.AspNet.OData.Shared/DeltaSetOfT.cs

```diff
@@ -0,0 +1,357 @@
+                            deltaSet.Add(changedObject);
+                            continue;
+                        }
+                        
+                        //Patch for addition/update. This will call Delta<T> for each item in the collection
+                        // This will work in case we use delegates for using users method to create an object
+                        changedObj.Patch(original, PatchHandler);                                                
+
+                        deltaSet.Add(changedObj);
+                    }
+                }
+                catch(Exception ex)
+                {
+                    //For handling the failed operations.
+                    IDeltaSetItem changedObject = HandleFailedOperation(changedObj, operation, original, ex.Message);                    
+                    deltaSet.Add(changedObject);
+                }
+            }
+
+            return deltaSet;
+        }
+
+
+        private DeltaSet<TStructuralType> CreateDetlaSet()
+        {
+            Type type = typeof(DeltaSet<>).MakeGenericType(_clrType);
+
+            DeltaSet<TStructuralType> deltaSet = Activator.CreateInstance(type, _keys) as DeltaSet<TStructuralType>;
+            return deltaSet;
+        }
+
+        private IDeltaSetItem HandleFailedOperation(Delta<TStructuralType> changedObj, DataModificationOperationKind operation, TStructuralType originalObj, string errorMessage)
+        {
+            IDeltaSetItem deltaSetItem = null;
+            DataModificationExceptionType dataModificationExceptionType = new DataModificationExceptionType(operation);
+            dataModificationExceptionType.MessageType = new MessageType();
+            dataModificationExceptionType.MessageType.Message = errorMessage;
+
+            // This handles the Data Modification exception. This adds Core.DataModificationException annotation and also copy other instance annotations.
+            //The failed operation will be based on the protocol
+            switch (operation)
+            {
+                case DataModificationOperationKind.Update:
+                    deltaSetItem = changedObj;
+                    break;
+                case DataModificationOperationKind.Insert:
+                    {
+                        deltaSetItem = CreateDeletedEntityForFailedOperation(changedObj);
+
+                        break;
+                    }
+                case DataModificationOperationKind.Delete:
+                    {
+                        deltaSetItem = CreateEntityObjectforFailedOperation(changedObj, originalObj);                        
+                        break;
+                    }
+            }
+
+
+            deltaSetItem.TransientInstanceAnnotationContainer = changedObj.TransientInstanceAnnotationContainer;
+            deltaSetItem.TransientInstanceAnnotationContainer.AddResourceAnnotation("Core.DataModificationException", dataModificationExceptionType);
```

Comment:
in the case of update, line 229 is basically a no-op (since deltaSetItem == changedObject) and line 230 adds the DataModificationException to the changedObj (which is probably okay, but we should think through to make sure changing changedObj is okay...)

Reply:
Its not always changed obj
in case of delete/insert